From 7dffafc1a9d7be561f1af24ffe934e34d42b792f Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <aaradhana.sahu@oss.qualcomm.com>
Date: Mon, 1 Sep 2025 14:06:20 +0530
Subject: [PATCH] UPSTREAM:CP: wifi: ath12k: Reduce power consumption during
 Wi-Fi down

Currently, when Wi-Fi goes down, power consumption remains high even if
no radio is active. To address this, the driver sends a WMI pdev_suspend
command to the firmware when the radio is inactive.

In the WSI bypass scenario, the SoC is bypassed if it's inactive, and the
pdev_suspend command is already sent in such cases. Remove the redundant
logic that attempts to send pdev_suspend again in the WSI bypass path.

Patch-dependency:
0441-03-UPSTREAM-CP-wifi-ath12k-Enable-pdev-suspend-and-resu.patch
Patch-work: None

Signed-off-by: Aaradhana Sahu <aaradhana.sahu@oss.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 32 +++++++++++++++++++++++++-
 drivers/net/wireless/ath/ath12k/mac.c  |  9 ++++----
 2 files changed, 35 insertions(+), 6 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 1ce5a2d..48db9eb 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -4284,13 +4284,43 @@ static void ath12k_core_wsi_remap_reset(struct ath12k_base *ab)
 
 int ath12k_core_dynamic_wsi_remap(struct ath12k_base *ab)
 {
-	int ret = 0;
+	int ret = 0, i;
 	struct ath12k_hw_group *ag;
+	struct ath12k *ar;
+	unsigned long time_left;
 
 	ag = ab->ag;
 
 	ag->wsi_remap_in_progress = true;
 
+	if (ab->wsi_remap_state == ATH12K_WSI_BYPASS_REMOVE_DEVICE) {
+		for (i = 0; i < ab->num_radios; i++) {
+			ar = ab->pdevs[i].ar;
+
+			if (ar && ar->pdev_suspend) {
+				reinit_completion(&ar->pdev_resume);
+				ret = ath12k_wmi_pdev_resume(ar, ar->pdev->pdev_id);
+
+				if (ret) {
+					ath12k_err(ar->ab,
+						   "failed to send wmi resume command %d\n",
+						   ret);
+					return ret;
+				}
+
+				time_left = wait_for_completion_timeout(&ar->pdev_resume,
+						ATH12K_PDEV_RESUME_TIMEOUT);
+
+				if (!time_left) {
+					ath12k_err(ar->ab,
+						   "Timeout in receiving pdev resume response: %d\n",
+						   ar->pdev->pdev_id);
+					return -ETIMEDOUT;
+				}
+			}
+		}
+	}
+
 	ath12k_dbg(ab, ATH12K_DBG_WSI_BYPASS, "WSI Bypass: MLO teardown with Umac reset");
 	ath12k_core_trigger_umac_reset(ab, WMI_MLO_TEARDOWN_REASON_DYNAMIC_WSI_REMAP);
 
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 2c61207..5502a76 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -14443,7 +14443,6 @@ int ath12k_mac_op_start(struct ieee80211_hw *hw)
 				ar->pdev_suspend = false;
 			}
 		}
-		ar->pdev_suspend = false;
 	}
 
 	if (ath12k_check_erp_power_down(ag))
@@ -15808,11 +15807,11 @@ err_vdev_del:
 	ahvif->device_bitmap &= ~BIT(ar->ab->wsi_info.index);
 
 	if (!ar->allocated_vdev_map && !arvif->is_scan_vif) {
-		if (ath12k_erp_get_sm_state() == ATH12K_ERP_ENTER_COMPLETE) {
-			ret = ath12k_mac_pdev_suspend(ar);
-			if (ret)
-				ath12k_warn(ab, "Pdev suspend command is failed %d\n", ret);
+		ret = ath12k_mac_pdev_suspend(ar);
+		if (ret)
+			ath12k_warn(ab, "Pdev suspend command is failed %d\n", ret);
 
+		if (ath12k_erp_get_sm_state() == ATH12K_ERP_ENTER_COMPLETE) {
 			if (ath12k_mac_validate_active_radio_count(ar->ah))
 				ath12k_core_cleanup_power_down_q6(ar->ah);
 		}
-- 
2.34.1

